### ISSUE TITLE
[$375] [HOLD for payment 2024-08-01] [Wave Collect] [Xero] Add 2FA requirement in Authorization flow

### ISSUE THREAD

**Issue by francoisl** [Expensify/App#43015]
### Problem
In order to remain compliant with Xero's third-party app requirements, we need to force workspace admins to enable 2FA before they can use the connection.

Additional internal context
- **[#whatsnext post](https://expensify.slack.com/archives/CC7NECV4L/p1715970581733459)**
- https://github.com/Expensify/Expensify/issues/397316

### Solution
- **For new connections:**
     - Before starting the OAuth flow, make sure the admin has 2FA enabled. If they don't, show the 2FA enablement page. After they finish setting it up, automatically resume the Xero auth flow.
     - See mockups in [this comment](https://github.com/Expensify/App/issues/43015#issuecomment-2150166743) below

- **For existing connections:**
    - If a user signs in and they're admin on a workspace that is connected to Xero but don't have 2FA enabled, show them the 2FA setup steps upon signing in
    - See mockups from [this comment](https://github.com/Expensify/App/issues/43015#issuecomment-2159166214)

cc @lakchote @zanyrenney 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f129776196b2ac5f</li>
        <li>Upwork Job ID: 1821257266596661328</li>
        <li>Last Price Increase: 2024-08-07</li>
    </ul>
</details>

**Comment by francoisl** [comment]
I'll add a few mockups when I have edit access to Figma again.

**Comment by dannymcclain** [comment]
@francoisl let me know if you need me to mock anything up or take care of anything in Figma! 🤝

**Comment by francoisl** [comment]
Thanks. I got an email from Figma saying that Jon gave me edit access but I still can't seem to edit anything.

Basically I was trying to edit the Initial Setup part of the [Xero mockups](https://www.figma.com/design/U9QcGPMLkoygY3HFkXU6eY/Wave-8%3A-Collect-Admins-Integrations?node-id=4964-13811&t=4qNOmzIQ0gXcZWoE-1) and add the 2FA flow in between. I was thinking of making two variations:
- one where we show the Two-factor authentication flow directly after you click Set up
- one with an intermediary modal with additional context, with text like "To connect to Xero, please enable Two-factor authentication first [Continue]"

**Comment by dannymcclain** [comment]
Super weird. I see that you were upgraded yesterday, but when I went into the file it still said you were requesting access (I gave it to you). I wonder if that file just hadn't been refreshed or something since Jon upgraded you. Anyways, try it again if you want, or I'm happy to do some jammin'.

**Comment by francoisl** [comment]
That's great, thanks Danny!

Looks like I was wrong about the sign-in flow and `<ValidateLoginPage>` doesn't do what I expected, so we'll also need to make some changes to make people enable 2FA when they sign in. I'm thinking we add an extra step after entering the magic code. Here's arough draft - same Figma link as Danny's above:

![image](https://github.com/Expensify/App/assets/2229301/6cc7fd79-893e-4e71-9d60-2a2c7703364c)

I changed the text description, but we can also make it more generic to accommodate the case when people sign in and they need to enable 2FA because they're on a domain group with that requirement.

**Comment by muttmuure** [comment]
Looks great to me

**Comment by rushatgabhane** [comment]
yo yo yo sorry i lost track of this issue. can you please assign it to me?

**Comment by rushatgabhane** [comment]
PR raised above ^

**Comment by rushatgabhane** [comment]
@francoisl everything tests well but we might have some hiccups on the backend.

For the flow: admin signs in and has a policy connected to Xero.

1. Login
2. Start 2FA steps
3. Scan QR code in authenticator
4. Paste the 6 digit code

Expected: the app becomes usable again.
Actual: the API call to enable 2FA completes but the app is unusable until we do a re-login.

https://github.com/Expensify/App/assets/29673073/5eba883f-9f03-41d1-a476-0efe6419f5ce

**Comment by francoisl** [comment]
Hm weird. I'll look into this today.

**Comment by francoisl** [comment]
As far as I can tell, it's because the authToken returned by `TwoFactorAuth_Validate` is not merged into Onyx before the next  API call to `ReconnectApp`.

In short, when 2FA is required to be enabled on sign-in like in our case here, the backend will first send an authToken that has very limited permissions - one of the very few API commands it can execute is `TwoFactorAuth_Validate`. That API command then returns a "regular" authToken that can execute other API commands.

As an example here, after signing in let's say I get this authToken `0322AAEA4169...` - this is a two-factor setup authToken with limited permissions.

<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/2229301/fa601922-3ea4-4488-83af-63a323e4dd20">

Once I enter an OTP code to enable 2FA, the client calls `TwoFactorAuth_Validate`, which returns a new authToken, in this case `E6CE5CDF6621...`

<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/2229301/b36ed821-d0ee-431a-a373-6088dadb6307">

However, immediately after we see that there is an API call to `ReconnectApp`, but for some reason it's still using the limited authToken `0322AAEA4169...`

<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/2229301/5e651f46-fb07-4e82-861e-a2da9d341b0c">

This is why the backend returns an error "Two Factor Authentication Required" and the app doesn't load. 

I _think_ the reason we call ReconnectApp before updating the authToken in Onyx is because of that `OnyxUpdates.saveUpdateInformation()` call [here](https://github.com/Expensify/App/blob/8a6ac298432d8c08a98d21f555affcd92c183699/src/libs/Middleware/SaveResponseInOnyx.ts#L32-L39), though I can't think of how to fix the issue. I tried adding `WRITE_COMMANDS.TWO_FACTOR_AUTH_VALIDATE` to [`const requestsToIgnoreLastUpdateID`](https://github.com/Expensify/App/blob/8a6ac298432d8c08a98d21f555affcd92c183699/src/libs/Middleware/SaveResponseInOnyx.ts#L6-L8) but that doesn't fully solve the issue because then we don't call `ReconnectApp` at all.

**Comment by rushatgabhane** [comment]
@francoisl thanks for looking into this! i think we'll have to convert `TwoFactorAuth_Validate` to `API_REQUEST_WITH_SIDE_EFFECTS`. We can then wait for the api call to complete and reconnect the app with the new token

**Comment by francoisl** [comment]
Final PR was just merged earlier today.

**Comment by c3024** [comment]
PR was reverted. 

@francoisl Could you please assign me this issue so that when the new PR is opened and marked ready puller bear correctly requests my review?

**Comment by sakluger** [comment]
I have a few payment-related questions:

1. @francoisl - this issue is labeled as Internal, but @rushatgabhane made one of the PRs (https://github.com/Expensify/App/pull/44059). Shouldn't this be External instead?
2. What should the price be on this one - $250 or $500?
3. It looks like there was a regression from this issue's PR. Do we need to halve the price?

**Comment by c3024** [comment]
@sakluger 

4. I reviewed the PR 😆

**Comment by rushatgabhane** [comment]
@sakluger i think 500 would be great because it was hard to nail down the nitty gritty of this task.

**Comment by rushatgabhane** [comment]
@c3024 could you please add regression test steps

**Comment by c3024** [comment]
## Regression test proposal

### Test 1:
#### Pre-requisite: 
Have an account `A` that 
      1. does not have 2FA.
      2. is an admin of a control policy `P`

#### Steps:
1. Login with the above specified account `A`
2. Click on Workspaces > Click on the control policy `P` specified in the pre-requisite > Click on Accounting in the left hand panel
3. Click on `Connect Xero`
4. Verify that 2FA modal is shown
5. Complete the 2FA setup flow
6. Verify that the Xero integration page opens now

### Test 2:
#### Pre-requisites: 
Have an account `A` that 
        1. does not have 2FA
        2. has a Xero accounting configured policy
        3. is an admin of this policy

#### Steps:
1. Login with the above specified account `A`
2. Verify that 2FA modal is shown with loading indicator behind
3. Click on `Enable Two Factor Authentication` button on the modal
4. Verify that the Avatar and Name in the left hand panel of Settings page has loading indicator
5. Click on `Two Factor Authentication` in the center panel and complete the 2FA setup flow
6. Verify that now the Avatar and Image in the left hand panel of Settings page is fully loaded with details
7. Click on `Inbox`
8. Verify that LHN and Center Pane load fully

**Comment by JmillsExpensify** [comment]
Can I get a payment summary on this issue?

**Comment by sakluger** [comment]
Sorry about that! I missed it because it wasn't automatically moved to Daily.

I chatted with @francoisl via DM to figure out what was going on here. We won't penalize for regressions here because the app was already unusable if you were an admin on a Xero workspace prior to the first PR.

Regarding price, we both thought that $500 was a bit high, but given the trickiness, we'll pay $375 - a bit more than the standard $250.

**Comment by sakluger** [comment]
@JmillsExpensify the payment summary is ready: https://github.com/Expensify/App/issues/43015#issuecomment-2263664335

**Comment by JmillsExpensify** [comment]
$375 approved for @rushatgabhane

**Comment by sakluger** [comment]
I completed the payment to @c3024 via Upwork.
